### PR TITLE
tslint, struct, tests

### LINKED ISSUE #1: Enable TypeScript's strict mode

### LINKED ISSUE #2: Setup testing

### LINKED ISSUE #4: Add tslint

### LINKED ISSUE #5: Setup CI

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes #1 
Fixes #2
Fixes #4 
Fixes #5

### CODE REVIEW

File: src/AttachAddon.test.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    server.close();
+  });
+
+  it('utf8', async function(): Promise<any> {
+    this.timeout(20000);
+    await openTerminal({ rendererType: 'dom' });
+    const port = 8080;
+    const server = new WebSocket.Server({ port });
+    const data = new Uint8Array([102, 111, 111]);
+    server.on('connection', socket => socket.send(data.buffer));
```
Comment (Tyriar) [inline src/AttachAddon.test.ts]: @jerch this doesn't seem to send it to the client when using buffers and I'm not sure why?

Comment (jerch) [reply src/AttachAddon.test.ts]: Seems ws' impl only supports the typed array, not the array buffer (see https://github.com/websockets/ws#sending-binary-data).

**Comment by jerch** [comment]
@Tyriar 
Furthermore I am not sure if there is a possible race condition hidden: The test sends the message as earlier as possible when `ws` got the connection ready. Now the question is if the browser will already consume messages between these two assignments as `Blob`:

```JS
const socket = new WebSocket(...)`;
socket.binarytype = 'arraybuffer';
```

The [spec](https://html.spec.whatwg.org/multipage/web-sockets.html#dom-websocket-binarytype) does not tell whether already buffered but not yet consumed messages will turn into array buffers by setting `binarytype`. Unfortunately the websocket ctor does not understand it as argument, ~~so there might be small time frame to miss a message due to still being Blob. Guess we have to test this.~~

Edit:
The second note in this [section](https://html.spec.whatwg.org/multipage/web-sockets.html#feedback-from-the-protocol) indicates, that those messages will be changed to typed arrays:
> User agents are encouraged to check if they can perform the above steps efficiently before they run the task, picking tasks from other task queues while they prepare the buffers if not. For example, if the binaryType attribute was set to "blob" when the data arrived, and the user agent spooled all the data to disk, but just before running the above task for this particular message the script switched binaryType to "arraybuffer", the user agent would want to page the data back to RAM before running this task so as to avoid stalling the main thread while it created the ArrayBuffer object.

**Comment by Tyriar** [comment]
Looks like it's still failing, and yeah I know about the race condition but didn't want to change much until I got it working. I verified earlier that send still didn't work (at least with .buffer) if I waited for a non-existant selector before checking the contents, this let's me see what happens in the browser.

**Comment by jerch** [comment]
The first message from server is a problem of its own. If I change the addon call from:
```TS
window.term.loadAddon(new window.AttachAddon(new WebSocket('ws://localhost:${port}')));
```

to

```TS
const ws = new WebSocket('ws://localhost:${port}');
const attachAddon = new window.AttachAddon(ws);
window.term.loadAddon(attachAddon);
```

I always miss the first message. This seems to be a known problem, but does not occur on all system (see https://stackoverflow.com/questions/34899954/prevent-theoretical-loss-of-the-first-received-message-in-websocket)

Solution: Wait on server side for an READY message from the attach addon - this also deals with the array buffer / blob problem above.

**Pull request merged**